### PR TITLE
chore: dependency reporting

### PR DESCRIPTION
**Pull request by mmanciop**
If the Lumigo OpenTelemetry distro for JS is configured to use the default Lumigo endpoint and has the `LUMIGO_TRACER_TOKEN` env var set, report the Node.js dependencies we can discover to the Lumigo backend for analysis using the `<DEFAULT_ENDPOINT>/v1/dependencies` endpoint to be provided by the Lumigo SaaS.

### CODE REVIEW

File: src/instrumentations/https/HttpInstrumentation.ts [hunk 1]
```diff
@@ -4,29 +4,37 @@ import { HttpHooks } from './http';
...
   }
 
   getInstrumentation(): HttpInstrumentation {
+    const ignoreConfig = this.getIgnoreConfig();
+
     return new HttpInstrumentation({
-      ignoreOutgoingUrls: this.getIgnoreConfig(),
-      ignoreIncomingPaths: this.getIgnoreConfig(),
+      ignoreOutgoingUrls: ignoreConfig,
+      ignoreIncomingPaths: [],
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/instrumentations/https/HttpInstrumentation.ts:28]: Change in the http instrumentation not to trace requests to the dependency endpoint

File: test/component/http/httpTestUtils.ts [hunk 2]
```diff
@@ -23,7 +23,6 @@ export const expectedResourceAttributes = {
         'process.environ': expect.jsonMatching(
             expect.objectContaining({
                 "OTEL_SERVICE_NAME": "http",
-                "LUMIGO_TRACER_TOKEN": "****",
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply test/component/http/httpTestUtils.ts:26]: That env var is not actually set

File: src/dependencies/report.test.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
+describe('dependency reporting', () => {
+  const ORIGINAL_PROCESS_ENV = process.env;
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/dependencies/report.test.js]: OK, dropping these tests, it's all covered in `wrapper.test.js` anyhow. Done in 485bb3c.

File: src/dependencies/report.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  lumigoToken: string,
+  resourceAttributes: ResourceAttributes
+) {
+  const packages = await listDependencies();
+
+  return postUri(
+    dependenciesEndpoint,
+    { resourceAttributes, packages },
+    { Authorization: `LumigoToken ${lumigoToken.trim()}` }
+  ).catch();
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/dependencies/report.ts:47]: Explicitly not. We do not want to raise questions for a best-effort analytics.

File: src/dependencies/report.test.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
+describe('dependency reporting', () => {
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply src/dependencies/report.test.js]: We already cover a few implicitly (our tests have no local `node_modules`. I am adding one for folders without `package.json` inside in 10ee4f8.

File: test/component/http/httpTestUtils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (shanishiri) [reply test/component/http/httpTestUtils.ts:26]: yes it is set, look at the only place where we use this - "basic http test" in http.test.js line 54

[the remark this answers is not in the document]
Comment (mmanciop) [reply test/component/http/httpTestUtils.ts:26]: It turns out it was a problem with the SIZE of the environment, as we truncate it to `20000` characters, and the build env has a huge process environment. Fixed in 4bcddc7.

File: src/instrumentations/https/HttpInstrumentation.ts [hunk 6]
```diff
@@ -4,29 +4,43 @@ import { HttpHooks } from './http';
...
     return [
-      (url: string) =>
-        [
-          process.env.LUMIGO_ENDPOINT,
-          process.env.ECS_CONTAINER_METADATA_URI,
-          process.env.ECS_CONTAINER_METADATA_URI_V4,
-        ]
-          .filter(Boolean)
-          .some((v) => url.includes(v)),
+      (url: string) => this.ignoredHostnames.includes(new URL(url).hostname),
```
Comment (shanishiri) [inline src/instrumentations/https/HttpInstrumentation.ts:42]: don't we need the "try-catch" also here?

File: src/wrapper.test.js [hunk 7]
```diff
@@ -179,220 +189,328 @@ describe('Distro initialization', () => {
...
-              throw new Error(`Unexpected url '${url}`);
-          }
-
-          return Promise.resolve(JSON.parse(responseRaw.toString()));
-        });
+        process.env.LUMIGO_TRACER_TOKEN = LUMIGO_TRACER_TOKEN;
+        process.env.OTEL_SERVICE_NAME = 'service-1';
       });
 
       test('NodeTracerProvider should be given a resource with all the right attributes', async () => {
```
Comment (shanishiri) [inline src/wrapper.test.js:282]: test with the same name - 3

File: src/wrapper.test.js [hunk 8]
```diff
@@ -179,220 +189,328 @@ describe('Distro initialization', () => {
...
   });
 
   describe('outside of a computing platform for which we have detectors', () => {
+    beforeEach(() => {
+      process.env.LUMIGO_TRACER_TOKEN = LUMIGO_TRACER_TOKEN;
+      process.env.OTEL_SERVICE_NAME = 'service-1';
+      process.env.LUMIGO_REPORT_DEPENDENCIES = 'false';
+    });
+
     test('NodeTracerProvider should be given a resource with all the right attributes', async () => {
```
Comment (shanishiri) [inline src/wrapper.test.js:210]: test with the same name repeat 4 times, although each is in another "describe" suite, can you pleas change the test name to be more precise?

File: src/wrapper.test.js [hunk 9]
```diff
@@ -179,220 +189,328 @@ describe('Distro initialization', () => {
...
-      test('NodeTracerProvider should be given a resource with all the right attributes', async () => {
-        jest.isolateModules(async () => {
-          process.env.LUMIGO_TRACER_TOKEN = LUMIGO_TRACER_TOKEN;
-          process.env.OTEL_SERVICE_NAME = 'service-1';
+      beforeEach(() => {
+        process.env.LUMIGO_TRACER_TOKEN = LUMIGO_TRACER_TOKEN;
+        process.env.OTEL_SERVICE_NAME = 'service-1';
+      });
 
+      test('NodeTracerProvider should be given a resource with all the right attributes', async () => {
```
Comment (shanishiri) [inline src/wrapper.test.js:237]: test with the same name - 2

File: src/wrapper.test.js [hunk 10]
```diff
@@ -179,220 +189,328 @@ describe('Distro initialization', () => {
...
-            detect: jest.fn().mockReturnValue(mockedResource),
-          },
-        };
-      });
+    beforeEach(() => {
+      process.env.LUMIGO_REPORT_DEPENDENCIES = 'false';
+    });
 
+    describe('on successful request', () => {
       test('NodeTracerProvider should be given a resource with all the right attributes', async () => {
```
Comment (shanishiri) [inline src/wrapper.test.js:328]: test with the same name - 4

File: src/wrapper.test.js [hunk 11]
```diff
@@ -1,5 +1,40 @@
 import * as fs from 'fs';
-import * as utils from './utils';
+
+/*
+ * Monkey-patch jest to be able to handle async in isolateModules
+ *
+ * See https://github.com/facebook/jest/issues/10428
+ */
+
+async function isolateModulesAsync(fn) {
```
Comment (shanishiri) [inline src/wrapper.test.js:9]: Nice workaround! can we put it in utils?

File: src/wrapper.test.js [hunk 12]
```diff
@@ -39,138 +74,113 @@ describe('Distro initialization', () => {
...
-      process.env.LUMIGO_TRACER_TOKEN = LUMIGO_TRACER_TOKEN;
+      /*
+       * We have a limit on the size of env we sent to the backend, and the env
+       * in the CI/CD goes over the limit, so the additional env vars we want to
+       * check for scrubbing get dropped.
+       */
+      process.env = {};
     });
 
-    test('should reject the init promise', () => {
```
Comment (shanishiri) [inline src/wrapper.test.js:66]: why those tests were removed?

**Review by shanishiri — APPROVED** [review]

File: src/wrapper.test.js (hunk 11 above)
Comment (mmanciop) [reply src/wrapper.test.js:9]: Feel free to, but not in this PR.

File: src/wrapper.test.js (hunk 10 above)
Comment (mmanciop) [reply src/wrapper.test.js:328]: It is not a problem, the hierarchy is different

File: src/wrapper.test.js (hunk 9 above)
Comment (mmanciop) [reply src/wrapper.test.js:237]: The hierarchy of the various levels of `describe` is different, no?

File: src/wrapper.test.js (hunk 8 above)
Comment (mmanciop) [reply src/wrapper.test.js:210]: What is the actual problem? The hierarchy matters, and that is how the failed tests are visualized: with the full hierarchy. Feel free to change them, but in a different PR.

File: src/instrumentations/https/HttpInstrumentation.ts (hunk 6 above)
Comment (mmanciop) [reply src/instrumentations/https/HttpInstrumentation.ts:42]: yeah I think you are right

Comment (mmanciop) [reply src/instrumentations/https/HttpInstrumentation.ts]: Done in c4ae63c

File: src/wrapper.test.js (hunk 12 above)
Comment (mmanciop) [reply src/wrapper.test.js:66]: Because they never worked. I found out by fixing the entire test harness.

**Pull request merged**